Source/Core/DolphinNoGUI/Platform.h

```diff
@@ -0,0 +1,48 @@
+// Copyright 2018 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "Common/Flag.h"
+#include "Common/WindowSystemInfo.h"
+
+class Platform
+{
+public:
+  virtual ~Platform();
+
+  bool IsRunning() const { return m_running.IsSet(); }
+  bool IsWindowFocused() const { return m_window_focus; }
+  bool IsWindowFullscreen() const { return m_window_fullscreen; }
+
+  virtual bool Init();
+  virtual void SetTitle(const std::string& title);
+  virtual void MainLoop() = 0;
+
+  virtual WindowSystemInfo GetWindowSystemInfo() const = 0;
+
+  // Requests a graceful shutdown, from SIGINT/SIGTERM.
+  void RequestShutdown();
+
+  // Request an immediate shutdown.
+  void Stop();
+
+  static std::unique_ptr<Platform> CreateHeadlessPlatform();
```

Comment:
I wonder if we should actually have those factory methods here, when they're only being used in a single place (that _also_ needs to be updated if another one is added).

Thoughts on just putting the `std::make_unique<$YourPlatformHere>()` call into `MainNoGUI.cpp` 's `GetPlatform`?
The only thing that might be a reason against this is providing a list somewhere that is also used to update the CLI help text that shows the platform selection (and even then the factory method itself has little use).

Reply:
The platform's class is declared in the .cpp file, so we'd have to move that to a header first. The main reason I'd want to avoid that would be pulling in the dependent headers (for example, the X11 headers define a lot of silly macros). Not sure which option is "cleaner".

Reply:
Ah, duh. Makes sense to hide the actual implementation detail.
Nvm this then.